text/1026-net-module.rst

```diff
@@ -0,0 +1,200 @@
+::
+
+    Status: Draft
+    Type: Feature
+    Created: 2024-08-14
+    Authors: Scott Trinh <scott@edgedb.com>
+
+=============================
+RFC: EdgeDB Networking Module
+=============================
+
+This RFC outlines the addition of a networking module to EdgeDB’s
+standard library, providing functionality for making HTTP requests and
+sending SMTP messages.
+
+Motivation
+==========
+
+The EdgeDB Auth extension (ext::auth) currently sends emails directly to
+configured SMTP services at certain times in various authentication
+ceremonies and workflows:
+
+-  Sign up with email (sends email verification email)
+-  Password reset (sends password reset email)
+-  Magic link sign up (sends magic link email)
+-  Magic link sign in (sends magic link email)
+
+To allow full customization of emails and provide hooks into
+authentication events generally, we need to support sending webhook
+requests to the application directly. This would enable applications to
+trigger their own logic for sending fully customized and
+internationalized emails, SMS, log analytics, update usage billing
+records, etc.
+
+Specification
+=============
+
+Module Name
+-----------
+
+The new module will be named ``net``.
+
+Functions
+---------
+
+The module will provide the following functions:
+
+1. HTTP Request Function:
+
+.. code-block:: edgeql
+
+   scalar type net::HttpMethod extending std::enums<`GET`, POST, PUT, `DELETE`, PATCH, HEAD, OPTIONS>;
+
+   function net::http_request(
+       url: str,
+       named only body: optional str,
+       named only method: net::HttpMethod = net::HttpMethod::GET,
+       named only headers: optional array<tuple<name: str, value: str>>
+   ) -> net::HttpResponse;
```

Comment:
> the task object links to both the request (for auditing purposes) and response object (more on the different states later)

Yeah, that's a good call. I would naturally reach for some kind of generic type in normal application code, so that makes sense.

> request_async

You mentioned adding a signifier like `async` in the original design meeting as well. My personal opinion is that _everything_ in this module should be async, so it feels weird to make the async-ness explicit. Seems to indicate that there is a non-async version as well. I'm not against adding it, but it feels noisy to me at the moment.